### PR TITLE
Add integration test for GunRequiresWieldComponent

### PR DESCRIPTION
**Pull request by VerinSenpai**
## About the PR
So it begins 😈 

Adds an integration test for the mosin. Honestly I don't interact with guns too often in game so I'm not entirely certain what all edge cases could cause a firearm to not fire besides not having ammo, not being racked, and not being wielded so please do grant me your guidance if you know of any I'm missing. Praying slarti doesn't baguette me. 🥖

Should be noted too that I did some testing to see what the best case was for checking the urist health. I devised that after spawning, its health should only be raising as it heals airloss. This means that the checks for damage should work fine so long as it doesn't start suffocating again 😠 I would like a second set of eyes to verify this though if you would be so kindly.

## Why / Balance
~~Prediction~~ Integration tests good.

## Technical details
Creates a new integration test in WeaponSniperTests.cs for the mosin.
Spawns a urist and puts a mosin in the players hand.
Applies a cooldown because its needed necessary.
For this test I'm asserting that GunRequiresWieldComp is present but that may not be necessary here and won't be used in any future tests.
Gets the current ammo count and wield status of the gun and verifies it has ammo and is not currently wielded.
Attempts to shoot the urist, which since the gun is not wielded, should fail.
Checks to see if any ammo has been spent or damage dealt. No? good.
Use in hand to wield the weapon and verify we've wielded it.
Gives the urist a taste of gat, mosin style.
Verify that we've only used one bullet and the urist has been damaged.

## Requirements

- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

### CODE REVIEW

File: Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
+﻿using Content.IntegrationTests.Tests.Interaction;
+using Content.Shared.Damage;
+using Content.Shared.Weapons.Ranged.Components;
+using Content.Shared.Weapons.Ranged.Events;
+using Content.Shared.Wieldable.Components;
+using Robust.Shared.Prototypes;
+
+namespace Content.IntegrationTests.Tests.Weapons;
+
+public sealed class WeaponSniperTests : InteractionTest
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs]:
```suggestion
public sealed class WeaponTests : InteractionTest
```
We will put other weapon tests in here later. Same for the file name.

File: Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+{
+    private static readonly EntProtoId MobHuman = "MobHuman";
+    private static readonly EntProtoId SniperMosin = "WeaponSniperMosin";
+
+    [Test]
+    public async Task SniperMosinInteractionTest()
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs]:
```suggestion
    public async Task GunRequiresWieldTest()
```
We are testing components, not this prototype in particular, so it would be better to name it after that.
Also leave a code comment what this is testing, like "Tests that a gun with GunRequiresWieldComponent can only be shot when wielded"

File: Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        var urist = await SpawnTarget(MobHuman);
+        var damageComp = Comp<DamageableComponent>(urist);
+        var startDamage = damageComp.TotalDamage.Value;
+
+        var mosinNet = await PlaceInHands(SniperMosin);
+        var mosinEnt = SEntMan.GetEntity(mosinNet);
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs]:
```suggestion
        var mosinEnt = ToServer(mosinNet);
```
you can use the shorthand for this

File: Content.IntegrationTests/Tests/Weapons/WeaponTests.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+{
+    private static readonly EntProtoId MobHuman = "MobHuman";
+    private static readonly EntProtoId SniperMosin = "WeaponSniperMosin";
+
+    [Test]
+    public async Task SniperMosinInteractionTest()
+    {
+        var urist = await SpawnTarget(MobHuman);
+        var damageComp = Comp<DamageableComponent>(urist);
+        var startDamage = damageComp.TotalDamage.Value;
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponTests.cs]:
```suggestion
        // The spawned mob starts with some airloss damage
        var startDamage = damageComp.TotalDamage.Value;
```
Was that the reason for this? If yes, leave it as a code comment.

File: Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+        await Pair.RunSeconds(2f); // Something about a cooldown or something IDK
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs]:
```suggestion
        await Pair.RunSeconds(2f); // Guns have a cooldown when picking them up.
```

File: Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+        Assert.That(HasComp<GunRequiresWieldComponent>(mosinNet),
+            "Looks like you've removed the 'GunRequiresWield' component from the mosin sniper." +
+            "If this was intentional, please update WeaponSniperTests.cs to reflect this change!");
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs]:
```suggestion
            "If this was intentional, please update GunRequiresWieldTest to reflect this change!");
```

File: Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+        var ammoEv = new GetAmmoCountEvent();
+        SEntMan.EventBus.RaiseLocalEvent(mosinEnt, ref ammoEv);
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs]: This should be really just be an API method somewhere in the SharedGunSystem that raises the event and returns the result. Like `GetAmmoCount` and `GetAmmoCapacity`. That way we don't have to directly raise the event here.

**Review by slarticodefast — CHANGES_REQUESTED** [review]
The test should fail with the default test mob by the way, since that only has one hand and cannot wield the mosin. But it turns out that is just broken and you can wield two-handed weapons when you only have one hand :death:
You can use the `removehand` command to test this. Looks like we will need another test for wielding with different amounts of hands.

grafik

**Comment by VerinSenpai** [comment]
You got it hoss. Figure after this, I can follow up with a test for hands and wielding that isnt just a side effect.

File: Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs (hunk 3 above)
Comment (VerinSenpai) [reply Content.IntegrationTests/Tests/Weapons/WeaponSniperTests.cs]: I didn't know that was what that was for. Neat. I was under the impression that was like dirty.

**Comment by VerinSenpai** [comment]
Still need to figure out how to update the player prototype with two hands. I'll handle that in a bit.

File: Content.IntegrationTests/Tests/Weapons/WeaponTests.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+using Content.Shared.Weapons.Ranged.Components;
+using Content.Shared.Weapons.Ranged.Systems;
+using Content.Shared.Wieldable.Components;
+using Robust.Shared.Prototypes;
+
+namespace Content.IntegrationTests.Tests.Weapons;
+
+public sealed class WeaponTests : InteractionTest
+{
+    private static readonly EntProtoId MobHuman = "MobHuman";
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponTests.cs:13]:
```suggestion
    protected override string PlayerPrototype => "MobHuman"; // The default test mob only has one hand
    private static readonly EntProtoId MobHuman = "MobHuman";
```
This way you can override the test mob and it will have two hands to use the guns with.
Without this the test will fail once we fix the bug.

File: Content.IntegrationTests/Tests/Weapons/WeaponTests.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    private static readonly EntProtoId SniperMosin = "WeaponSniperMosin";
+
+    [Test]
+    public async Task GunRequiresWieldTest()
+    {
+        var gunSystem = SEntMan.System<SharedGunSystem>();
+
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponTests.cs:20]:
```suggestion

        await AddAtmosphere(); // prevent the Urist from suffocating
```

File: Content.IntegrationTests/Tests/Weapons/WeaponTests.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            "Looks like you've removed the 'GunRequiresWield' component from the mosin sniper." +
+            "If this was intentional, please update WeaponTests.cs to reflect this change!");
+
+        var startAmmo = gunSystem.GetAmmoCount(mosinEnt);
+        var wieldComp = Comp<WieldableComponent>(mosinNet);
+
+        Assert.That(startAmmo, Is.GreaterThan(0), "Mosin was spawned with no ammo!");
+        Assert.That(wieldComp.Wielded, Is.False, "Mosin was spawned wielded!");
+
+        await AttemptShoot(urist, false);
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponTests.cs]:
```suggestion
        await AttemptShoot(urist, false); // should fail due to not being wielded
```

File: Content.IntegrationTests/Tests/Weapons/WeaponTests.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        var updatedAmmo = gunSystem.GetAmmoCount(mosinEnt);
+
+        Assert.That(updatedAmmo,
+            Is.EqualTo(startAmmo),
+            "Mosin discharged ammo when the weapon should not have fired!");
+        Assert.That(damageComp.TotalDamage.Value,
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponTests.cs:47]: Wait, you are asserting that they took damage here, likely from suffocating.

**Review by slarticodefast — CHANGES_REQUESTED** [review]

File: Content.IntegrationTests/Tests/Weapons/WeaponTests.cs (hunk 8 above)
Comment (VerinSenpai) [reply Content.IntegrationTests/Tests/Weapons/WeaponTests.cs:13]: whoa, wicked.

File: Content.IntegrationTests/Tests/Weapons/WeaponTests.cs (hunk 11 above)
Comment (VerinSenpai) [reply Content.IntegrationTests/Tests/Weapons/WeaponTests.cs:47]: Right, I was checking to verify that the urist *didnt* take damage when the gun should not have discharged.

Comment (VerinSenpai) [reply Content.IntegrationTests/Tests/Weapons/WeaponTests.cs:47]: *tsh* oh, I see, I'm a dunce.

File: Content.IntegrationTests/Tests/Weapons/WeaponTests.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    [Test]
+    public async Task GunRequiresWieldTest()
+    {
+        var gunSystem = SEntMan.System<SharedGunSystem>();
+
+        await AddAtmosphere(); // prevent the Urist from suffocating
+
+        var urist = await SpawnTarget(MobHuman);
+        var damageComp = Comp<DamageableComponent>(urist);
+        // The spawned mob starts with some air loss damage.
```
Comment (slarticodefast) [inline Content.IntegrationTests/Tests/Weapons/WeaponTests.cs]:
```suggestion
```
The comment is redundant now.

**Comment by VerinSenpai** [comment]
Screenshot_20251021-122620

I held the power of the world in my hands, if only for a moment.
